course/src/isometry.cpp

```diff
@@ -180,5 +180,181 @@ namespace math {
+  }
+
+  Matrix3 Matrix3::operator/(const double divider) const {
+    return {
+      rows.at(0)/divider,
+      rows.at(1)/divider,
+      rows.at(2)/divider};
+  }
+
+  const Matrix3 operator*(const Matrix3& matrix1, const double scalar) {
+    return {
+      matrix1.rows.at(0)*scalar,
+      matrix1.rows.at(1)*scalar,
+      matrix1.rows.at(2)*scalar};
+  }
+
+  const Matrix3 operator*(const double scalar, const Matrix3& matrix1) {
+    return {
+      matrix1.rows.at(0)*scalar,
+      matrix1.rows.at(1)*scalar,
+      matrix1.rows.at(2)*scalar};
+  }
+
+  double Matrix3::det() const {
+    return
+      rows[0][0] * (rows[1][1]*rows[2][2] - rows[1][2]*rows[2][1]) -
+      rows[0][1] * (rows[1][0]*rows[2][2] - rows[1][2]*rows[2][0]) +
+      rows[0][2] * (rows[1][0]*rows[2][1] - rows[1][1]*rows[2][0]);
+  }
+
+  const Vector3 operator*(const Matrix3& matrix1, const Vector3& vector1) {
+    return Vector3(
+      matrix1.rows.at(0).dot(vector1),
+      matrix1.rows.at(1).dot(vector1),
+      matrix1.rows.at(2).dot(vector1));
+  }
+
+  Matrix3& Matrix3::operator+=(const Matrix3& matrix1) {
+    rows.at(0) += matrix1.rows.at(0);
+    rows.at(1) += matrix1.rows.at(1);
+    rows.at(2) += matrix1.rows.at(2);
+    return *this;
+  }
+
+  Matrix3& Matrix3::operator-=(const Matrix3& matrix1) {
+    rows.at(0) -= matrix1.rows.at(0);
+    rows.at(1) -= matrix1.rows.at(1);
+    rows.at(2) -= matrix1.rows.at(2);
+    return *this;
+  }
+
+  Matrix3& Matrix3::operator*=(const Matrix3& matrix1) {
+    rows.at(0) *= matrix1.rows.at(0);
+    rows.at(1) *= matrix1.rows.at(1);
+    rows.at(2) *= matrix1.rows.at(2);
+    return *this;
+  }
+
+  Matrix3& Matrix3::operator/=(const Matrix3& matrix1) {
+    rows.at(0) /= matrix1.rows.at(0);
+    rows.at(1) /= matrix1.rows.at(1);
+    rows.at(2) /= matrix1.rows.at(2);
+    return *this;
+  }
+
+  Matrix3& Matrix3::operator*=(const double scalar) {
+    rows.at(0) *= scalar;
+    rows.at(1) *= scalar;
+    rows.at(2) *= scalar;
+    return *this;
+  }
+
+  Matrix3& Matrix3::operator/=(const double scalar) {
+    rows.at(0) /= scalar;
+    rows.at(1) /= scalar;
+    rows.at(2) /= scalar;
+    return *this;
+  }
+
+  Vector3 Matrix3::operator[](int row) const {
+    return rows.at(row);
+  }
+
+  Vector3 & Matrix3::operator[](int row) {
+    return rows.at(row);
+  }
+
+  std::ostream& operator<<(std::ostream &ss, const Matrix3& matrix1) {
+    ss << "[["
+       << matrix1[0][0] << ", "
+       << matrix1[0][1] << ", "
+       << matrix1[0][2] << "], ["
+
+       << matrix1[1][0] << ", "
+       << matrix1[1][1] << ", "
+       << matrix1[1][2]  << "], ["
+       << matrix1[2][0] << ", "
+       << matrix1[2][1] << ", "
+       << matrix1[2][2]
+       << "]]";
+    return ss;
+  }
+
+  Vector3 Matrix3::col(const int column) const {
+    return { rows.at(0)[column], rows.at(1)[column], rows.at(2)[column]};
+  }
+
+  Vector3 Matrix3::row(const int row) const {
+    return { rows.at(row)};
```

Comment:
nit: remove the `{ ... }` wrapping, since the `at()` method already returns a vector, it's not needed, and might cause one more temporary object to be created if [copy elision](https://en.cppreference.com/w/cpp/language/copy_elision) were off or unable to detect this.

Reply:
See d932b3f